File: golang/middleware.go
package main

import (
	"crypto/sha256"
	"crypto/subtle"
	"encoding/hex"
	"fmt"
	"io"
	"log"
	"net/http"

	"github.com/golang-jwt/jwt"
)

func withWebhookSignature(secret, issuer string, next http.HandlerFunc) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		signature := r.Header.Get("X-Webhook-Signature")
		if signature == "" {
			log.Println("❌ X-Webhook-Signature header not present")
			http.Error(w, "Unauthorized", http.StatusUnauthorized)
			return
		}

		body, err := io.ReadAll(r.Body)
		if err != nil {
			log.Printf("❌ Error reading request body: %v", err)
			http.Error(w, err.Error(), http.StatusInternalServerError)
			return
		}
		defer r.Body.Close()

		sha := sha256.Sum256(body)
		bodyHash := hex.EncodeToString(sha[:])

		token, err := jwt.Parse(signature, func(token *jwt.Token) (interface{}, error) {
			if _, ok := token.Method.(*jwt.SigningMethodHMAC); !ok {
				return nil, fmt.Errorf("Unexpected signing method: %v", token.Header["alg"])
			}
			if token.Header["alg"] != "HS256" {
				return nil, fmt.Errorf("Unexpected signing algorithm: %v", token.Header["alg"])
			}
			return []byte(secret), nil
		})

		if claims, ok := token.Claims.(jwt.MapClaims); ok && token.Valid {
			if claims["iss"] != issuer {
				log.Printf("❌ Incorrect issuer: %q", claims["iss"])
				http.Error(w, "Invalid JWT issuer", http.StatusForbidden)
				return
			}

			if 1 != subtle.ConstantTimeCompare([]byte(claims["sha256"].(string)), []byte(bodyHash)) {
				log.Println("❌ Body SHA256 does not match")
				log.Printf("   JWT: %q", claims["sha256"])
				log.Printf("Actual: %q", bodyHash)
				http.Error(w, "Invalid JWT body hash", http.StatusForbidden)
				return
			}
		} else {
			log.Printf("❌ Could not decode JWT: %v", err)
			http.Error(w, "Invalid JWT", http.StatusForbidden)
			return
		}

		log.Println("✅ X-Webhook-Signature Valid");

		next.ServeHTTP(w, r)
	}
}

File: nodejs/middleware.js
import jwt from 'jwt-simple';
import { createHash, timingSafeEqual } from 'crypto';

export default function WebhookSignatureMiddleware (secret, issuer) {
  return (req, res, next) => {
    const header = req.header('X-Webhook-Signature')
    if(!header) {
      console.log('❌ X-Webhook-Signature header not present');
      res.status(401).send('Unauthorized');
      return;
    }

    let decoded;
    let expectedSha256;

    try {
      decoded = jwt.decode(header, secret, false, 'HS256')
      expectedSha256 = createHash('sha256').update(req.body).digest();
    } catch(err) {
      console.log(`❌ Could not decode JWT: ${err}`);
      res.status(403).send('Invalid JWT');
      return;
    }

    if(decoded.iss != issuer) {
      console.error(`❌ Incorrect issuer: ${decoded.iss}`);
      res.status(403).send('Invalid JWT issuer');
      return;
    }

    if(!timingSafeEqual(Buffer.from(decoded.sha256, 'hex'), expectedSha256)) {
      console.log('❌ Body SHA256 does not match');
      console.log(`   JWT: ${decoded.sha256}`);
      console.log(`Actual: ${expectedSha256.toString('hex')}`);
      res.status(403).send('Invalid JWT body hash');
      return;
    }

    console.log('✅ X-Webhook-Signature Valid');

    next();
  }
}
